File: windows/GitHubSignal.Core/PullRequestInfo.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace GitHubSignal.Core;

public sealed record PullRequestInfo(string? Status, DateTimeOffset CheckedAt)
{
    [System.Text.Json.Serialization.JsonIgnore]
    public string Label => Status switch { "open" => "🟢 Open", "draft" => "⚪ Draft", "merged" => "🟣 マージ済み", "closed" => "🔴 クローズ", _ => "状態未取得" };
    public static PullRequestInfo FromJson(JsonElement value) {
        var state = value.TryGetProperty("state", out var s) ? s.GetString() : null;
        bool merged = value.TryGetProperty("merged", out var m) && m.ValueKind == JsonValueKind.True;
        bool draft = value.TryGetProperty("draft", out var d) && d.ValueKind == JsonValueKind.True;
        return new(merged ? "merged" : state == "closed" ? "closed" : state == "open" ? (draft ? "draft" : "open") : null, DateTimeOffset.UtcNow);
    }
    public static string? ApiPath(Signal signal) {
        if (Rules.SafeWebUrl(signal.Url) is not { } url || url.Host != "github.com") return null;
        var match = Regex.Match(url.AbsolutePath, @"^/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+/pull/[0-9]+(?=/|$)");
        if (!match.Success) return null;
        var parts = match.Value.Split('/');
        return $"/repos/{parts[1]}/{parts[2]}/pulls/{parts[4]}";
    }
}

public sealed record SignalBatch(List<Signal> Signals, string? ThreadKey = null, PullRequestInfo? PullRequest = null);

File: Sources/NotificationCore/PullRequestInfo.swift
import Foundation

public struct PullRequestInfo: Codable, Equatable {
    public var status: String?
    public var checkedAt: Date

    public init(state: String?, draft: Bool?, merged: Bool?, checkedAt: Date = Date()) {
        status = merged == true ? "merged" : state == "closed" ? "closed" : state == "open" ? (draft == true ? "draft" : "open") : nil
        self.checkedAt = checkedAt
    }

    public var label: String {
        switch status {
        case "open": return "🟢 Open"
        case "draft": return "⚪ Draft"
        case "merged": return "🟣 マージ済み"
        case "closed": return "🔴 クローズ"
        default: return "状態未取得"
        }
    }

    public static func apiPath(for signal: Signal) -> String? {
        guard let url = SignalRules.safeWebURL(signal.url), url.host == "github.com",
              let match = url.path.range(of: #"^/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+/pull/[0-9]+(?=/|$)"#, options: .regularExpression) else { return nil }
        let parts = url.path[match].split(separator: "/")
        return "/repos/\(parts[0])/\(parts[1])/pulls/\(parts[3])"
    }
}

public struct SignalBatch {
    public var signals: [Signal]
    public var threadKey: String?
    public var pullRequest: PullRequestInfo?
}
